aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,158 @@
+	- Save the business entity to persistence
+	- Return an appropriate ``IActionResult``
+
+Unit Testing
+------------
+:doc:`Unit testing </testing/unit-testing>` involves testing a part of an app in isolation from its infrastructure and dependencies. When unit testing controller logic, only the contents of a single action should be tested, not the behavior of its dependencies or of the framework itself. Thus, when unit testing a controller's actions, avoid testing global or attribute-based :doc:`filters <filters>`, :doc:`routing </fundamentals/routing>`, or :doc:`model binding </mvc/models/model-binding>`, as these are all performed by the framework (testing these is an integration test responsibility).
+
+To demonstrate unit testing, review the following controller. It displays a list of brainstorming sessions and allows new brainstorming sessions to be created with a POST:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Controllers/HomeController.cs
+  :language: c#
+  :emphasize-lines: 11,15,20,40-41
+
+The controller is following the `explicit dependencies principle <http://deviq.com/explicit-dependencies-principle/>`_, expecting dependency injection to provide it with an instance of ``IBrainStormSessionRepository``. This makes it fairly easy to test using a mock object framework, like `Moq <https://www.nuget.org/packages/Moq/>`_. To test the first ``Index`` method, which has no looping or branching logic and only calls one method, we need only verify that a ``ViewResult`` is returned, with a ``ViewModel`` containing whatever was returned from the repository's ``List`` method.
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/HomeControllerIndex.cs
+  :language: c#
+  :emphasize-lines: 16-17,23-25
+
+The second ``Index`` method (which accepts an ``HttpPost``) has slightly more logic, since it checks ``ModelState.IsValid``. We should test that, when ``ModelState.IsValid`` is ``false`` the action method returns a ``ViewResult`` with the appropriate data. Otherwise, we should confirm that the ``Add`` method is called on the repository and a ``RedirectToActionResult`` is returned with the correct arguments. Below are its unit tests:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/HomeControllerIndexPost.cs
+  :language: c#
+  :emphasize-lines: 14-15,20,27-28,31,38
+
+The first test confirms that, when the ``ModelState`` is not valid, the method returns the same ``ViewResult`` as a GET request would. Note that the test doesn't attempt to pass in an invalid model. That wouldn't work anyway since model binding isn't running - we're just calling the method directly. However, we're not trying to test model binding - we're only testing what our code in the action method does. So the simplest thing is to just explicitly add an error to ``ModelState``.
```

Comment:
drop "that," or at least comma
The first test confirms when ``ModelState`` is not valid, the same ``ViewResult`` is returned as a GET request.